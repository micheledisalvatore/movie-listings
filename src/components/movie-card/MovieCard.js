import React from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'ramda';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';
import Chip from '@material-ui/core/Chip';

export const MovieCard = ({ classes, image, title, genreIds, genres }) => (
  <Card className={classes.card}>
    <div className={classes.details}>
      <CardMedia
        component="img"
        alt="Contemplative Reptile"
        className={classes.media}
        image={`https://image.tmdb.org/t/p/w500/${image}`}
        title={title}
      />
      <CardContent>
        <Typography gutterBottom variant="h5" component="h2">
          {title}
        </Typography>
        <div>
        {!isEmpty(genres) && genreIds.map(genreId => {
          const { name } = genres[genreId];
          return (
            <Chip
              key={genreId}
              label={name}
              className={classes.chip}
            />
          )}
        )}
        </div>
      </CardContent>
    </div>
  </Card>
);

MovieCard.propTypes = {
  classes: PropTypes.object.isRequired,
  genreIds: PropTypes.arrayOf(PropTypes.number),
  genres: PropTypes.shape({
    name: PropTypes.string,
  }),
  image: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
};

MovieCard.defaultProps = {
  genreIds: [],
  genres: {},
}

export const mapStateToProps = ({ genres: { list }}) => ({
  genres: list || {},
})