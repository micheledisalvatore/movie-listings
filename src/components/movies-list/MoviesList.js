import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { values } from 'ramda';

import Grid from '@material-ui/core/Grid';

import MovieCard from '../movie-card';
import { getMovies } from '../../actions/movies';

export class MoviesList extends Component {
  constructor(props) {
    super(props);

    props.getMoviesAction();
  }

  render() {
    const { classes, movies } = this.props;
    return (
      <div className={classes.root}>
        <Grid container spacing={24}>
          { movies.map(({
            id, poster_path: image, title, genre_ids: genreIds,
          }) => (
            <Grid item xs={12} sm={6} md={3} key={id} className={classes.item}>
              <MovieCard image={image} title={title} genreIds={genreIds} />
            </Grid>
          ))}
        </Grid>
      </div>
    );
  }
}

MoviesList.propTypes = {
  classes: PropTypes.object.isRequired,
  getMoviesAction: PropTypes.func.isRequired,
  movies: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      genre_ids: PropTypes.arrayOf(PropTypes.number),
      poster_path: PropTypes.string,
      title: PropTypes.string,
    }),
  ),
};

MoviesList.defaultProps = {
  movies: [],
};

const filterMovies = (filterRating, filterGenres) => ({
  vote_average: voteAverage,
  genre_ids: genreIds,
}) => {
  const isMatchingRating = voteAverage >= (filterRating * 2);
  const hasSelectedCategory = filterGenres.every(filterGenre => genreIds.indexOf(filterGenre) > -1);

  return isMatchingRating && (filterGenres.length === 0 || hasSelectedCategory);
};

const sortByPopularity = (a, b) => b.popularity - a.popularity;

export const mapStateToProps = ({ movies: { list, filterRating }, genres: { filters } }) => ({
  movies: values(list).filter(filterMovies(filterRating, filters)).sort(sortByPopularity) || [],
});

export const mapDispatchToProps = dispatch => ({
  getMoviesAction: bindActionCreators(getMovies, dispatch),
});
