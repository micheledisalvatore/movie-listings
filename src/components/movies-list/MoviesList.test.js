/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import Grid from '@material-ui/core/Grid';

import MovieCard from '../movie-card';
import { MoviesList, mapStateToProps } from './MoviesList';

describe('Given a MoviesList component', () => {
  let component;
  let wrapper;
  const getMoviesActionMock = jest.fn();
  const requiredProps = {
    classes: {},
    getMoviesAction: getMoviesActionMock,
  };

  describe('when it is rendered', () => {
    beforeEach(() => {
      component = <MoviesList {...requiredProps} />;
      wrapper = shallow(component);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should NOT render any MovieCard', () => {
      expect(wrapper.find(MovieCard)).toHaveLength(0);
    });

    it('should call the getMoviesAction', () => {
      expect(getMoviesActionMock).toHaveBeenCalledWith();
    })

    describe('and the list of movies is populated', () => {
      beforeEach(() => {
        wrapper.setProps({
          movies: [{
            id: 0,
            genre_ids: [1,2,3],
            poster_path: 'fooUrl',
            title: 'fooTitle',
          }]
        })
      });

      it('should render one MovieCard item', () => {
        expect(wrapper.find(MovieCard)).toHaveLength(1);
      });

      it('should render a MovieCard with the given title', () => {
        expect(wrapper.find(MovieCard)).toHaveProp('title', 'fooTitle');
      });

      it('should render a MovieCard with the given genreIds', () => {
        expect(wrapper.find(MovieCard)).toHaveProp('genreIds', [1,2,3]);
      });

      it('should render a MovieCard with the given image', () => {
        expect(wrapper.find(MovieCard)).toHaveProp('image', 'fooUrl');
      });
    })
  });
});

describe('Given a mapStateToProps function', () => {
  describe('when it is called with an empty filterGenres', () => {
    it('should return a sorted object with all the passed movies', () => {
      expect(mapStateToProps({
        movies: {
          list: {
            0: {
              vote_average: 1,
              genre_ids: [0,1],
              popularity: 3,
            },
            1: {
              vote_average: 1,
              genre_ids: [1,3],
              popularity: 5,
            },
          },
          filterRating: 0,
        },
        genres: {
          filters: [],
        },
      })).toEqual({
        movies: [
          {
            vote_average: 1,
            genre_ids: [1,3],
            popularity: 5,
          },{
            vote_average: 1,
            genre_ids: [0,1],
            popularity: 3,
          }
        ],
      });
    });

    it('should return an object with all the passed movies', () => {
      expect(mapStateToProps({
        movies: {
          list: {
            0: {
              vote_average: 1,
              genre_ids: [0,1],
              popularity: 5,
            },
          },
          filterRating: 0,
        },
        genres: {
          filters: [],
        },
      })).toEqual({
        movies: [
          {
            vote_average: 1,
            genre_ids: [0,1],
            popularity: 5,
          }
        ],
      });
    });
  });

  describe('when it is called with a filled filterGenres', () => {
    it('should return a filtered object with no matching movies', () => {
      expect(mapStateToProps({
        movies: {
          list: {
            0: {
              vote_average: 1,
              genre_ids: [0,1],
              popularity: 3,
            },
            1: {
              vote_average: 1,
              genre_ids: [1,3],
              popularity: 5,
            },
            3: {
              vote_average: 3,
              genre_ids: [0,3],
              popularity: 6,
            },
          },
          filterRating: 0,
        },
        genres: {
          filters: [3, 2],
        },
      })).toEqual({
        movies: [],
      });
    });

    it('should return a filtered object with 1 matching movie', () => {
      expect(mapStateToProps({
        movies: {
          list: {
            0: {
              vote_average: 1,
              genre_ids: [0,1],
              popularity: 3,
            },
            1: {
              vote_average: 1,
              genre_ids: [1,3],
              popularity: 5,
            },
            3: {
              vote_average: 3,
              genre_ids: [0,3],
              popularity: 6,
            },
          },
          filterRating: 0,
        },
        genres: {
          filters: [3, 1],
        },
      })).toEqual({
        movies: [{
          vote_average: 1,
          genre_ids: [1,3],
          popularity: 5,
        }],
      });
    });

    it('should return a filtered object with no matching movies when rating is not satisfied', () => {
      expect(mapStateToProps({
        movies: {
          list: {
            0: {
              vote_average: 1,
              genre_ids: [0,1],
              popularity: 3,
            },
            1: {
              vote_average: 1,
              genre_ids: [1,3],
              popularity: 5,
            },
            3: {
              vote_average: 3,
              genre_ids: [0,3],
              popularity: 6,
            },
          },
          filterRating: 3,
        },
        genres: {
          filters: [3, 1],
        },
      })).toEqual({
        movies: [],
      });
    });
  });
});
