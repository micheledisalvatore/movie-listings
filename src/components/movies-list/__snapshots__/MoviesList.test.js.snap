// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a MoviesList component when it is rendered should match the snapshot 1`] = `
<div>
  <WithStyles(Grid)
    container={true}
    spacing={24}
  />
</div>
`;
